### ISSUE TITLE
Fail to build in CPU only environment

### ISSUE THREAD

**Issue by cloudtoro** [smallcorgi/Faster-RCNN_TF#24]
Hi,

Errors were encountered when building Faster-RCNN-TF in CPU only environment. 

It still did not succeed even after all "cuda" references were removed from setup.py.

Any suggestions?

Thanks!
cloud

**Comment by ahmedammar** [comment]
I have a WIP attempt of CPU only, I'm testing the training now and will probably push it later. The main point is that there is a missing CPU implementation for RoiPoolGrad. I still need to work clean it up so it works with either gpu or cpu, right now I have gpu completely disabled.

**Comment by ljyyoruhi** [comment]
@ahmedammar Look forward to your CPU only version. It may help me a lot! Sorry to bother you, but, I can't wait until the new branch pulled. So may I have some suggestion? I have done some modifications like the following:

1. In ./lib/fast_rcnn/config.py, set "USE_GPU_NMS" to False.

2. In ./tools/test_net.py / train_net.py, replace "device_name = '/gpu:{:d}'.format(args.gpu_id)" with "device_name = '/cpu:0'"

3. In ./lib/setup.py, I comment these lines 
...
#CUDA = locate_cuda()
...
#self.set_executable('compiler_so', CUDA['nvcc'])
...
#Extension('nms.gpu_nms',
# ['nms/nms_kernel.cu', 'nms/gpu_nms.pyx'],
# library_dirs=[CUDA['lib64']],
# libraries=['cudart'],
# language='c++',
# runtime_library_dirs=[CUDA['lib64']],
# # this syntax is specific to this build system
# # we're only going to use certain compiler args with nvcc and not with
# # gcc the implementation of this trick is in customize_compiler() below
# extra_compile_args={'gcc': ["-Wno-unused-function"],
# 'nvcc': ['-arch=sm_35',
# '--ptxas-options=-v',
# '-c',
# '--compiler-options',
# "'-fPIC'"]},
# include_dirs = [numpy_include, CUDA['include']]
# )

4. In ./lib/fast_rcnn/nms_wrapper.py, I comment
...
#from nms.gpu_nms import gpu_nms
...
# if cfg.USE_GPU_NMS and not force_cpu:
        # return gpu_nms(dets, thresh, device_id=cfg.GPU_ID)
    # else:
return cpu_nms(dets, thresh)

5. Finally, I also do "sudo pip install --upgrade https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-0.8.0-cp27-none-linux_x86_64.whl" to make TensorFlow work on CPU only mode.

6. Then I do "make" and run "demo.py". But I still get a traceback at line 117 in demo.py, and get the error of tensorflow.python.framework.errors.PermissionDeniedError: /home/ubuntu/.nv

Have you also got this error before? And did I miss anything? Appreciate if you can have suggestion on it. Thank you very much!

**Comment by ahmedammar** [comment]
@ljyyoruhi I already committed my cpu-only changes .. here #28

**Comment by ljyyoruhi** [comment]
@ahmedammar Got it! Thank you very much!

**Comment by zhaotiny** [comment]
@ljyyoruhi Did you run CPU-only mode successfully?

**Comment by ljyyoruhi** [comment]
@zhaotiny Yes, the demo can run successfully. But for there will be some problems with the subplots function in demo.py when running demo on remote machine, I comment those related code.

**Comment by emily2he** [comment]
@ahmedammar  I only want to build cpu demo,so I comment everything about gpu in lib/make.sh. tensorflow version is only cpu 1.1
when I run make.sh, it report error:roi_pooling_op.cc:25:47: fatal error: tensorflow/core/util/work_sharder.h: No such file or directory
I have no idea how to solve this problem, please help me, I will appreciate that.

**Comment by cloudtoro** [comment]
Can it run with the CPU-only version of TensorFlow?

Thanks,
cloud

**Comment by BW01** [comment]
I get the same error as @emily2he when building on cpu.

**Comment by howardlei82** [comment]
@ljyyoruhi Can you share the environment you used to run the demo in CPU-mode only? I'm still having trouble running on CPU-only mode.

Thanks,
-Howard
